### ISSUE TITLE
Extremely small validation loss

### ISSUE THREAD

**Issue by SeaBroNoMeal** [fizyr/keras-retinanet#1385]
Hello, I am running retinanet on 16bit image, I replaced the preprocessing part by mapping my range to [-1.0, 1,0] (float number), and during training I am getting this result like:

`3000/3000 [==============================] - 2486s 829ms/step - loss: 1.2900 - regression_loss: 0.6755 - classification_loss: 0.6145 - val_loss: 7.9232e-06 - val_regression_loss: 0.6628 - val_classification_loss: 0.4772`

I think the training loss is acceptable and reasonable, but the validation loss was strange (so small). And at the same time, my mAP is always zero, is there any common mistake that could lead to this behaviour? And is there any other codes I have to modify? I only change the` preprocess_image `function in the `image.py` .
Thanks in advance!

**Comment by yashmarathe21** [comment]
I came across a similar problem.
Have you found the solution to this?

**Comment by yashmarathe21** [comment]
I couldn't find the bug. But the issue is only with validation loss. So, it won't affect the training.
But you can just add a callback in which you add the val_regression_loss and val_classification_loss and print that. 

class LossHistory(keras.callbacks.Callback):
    
    def __init__(self, verbose):
        self.verbose = verbose
        super(LossHistory, self).__init__()
    
    def on_train_begin(self, logs={}):
        self.history = {'loss':[], 'regression_loss':[], 'classification_loss':[],
                        'total_val_loss':[], 'val_regression_loss':[], 'val_classification_loss':[]}
        
    def on_epoch_end(self, epoch, logs={}):
        
        
        print("Epoch - ",epoch+1)
        
        self.history['regression_loss'].append(logs.get('regression_loss'))
        self.history['classification_loss'].append(logs.get('classification_loss'))
        self.history['loss'].append(logs.get('loss'))
        
        self.history['val_regression_loss'].append(logs.get('val_regression_loss'))
        self.history['val_classification_loss'].append(logs.get('val_classification_loss'))
        self.history['total_val_loss'].append(logs.get('val_regression_loss')+logs.get('val_classification_loss'))
        
        logs['total_val_loss'] = self.history['total_val_loss'][-1]
        
        print('train loss = ',logs.get('loss'))
        print('train regression loss = ',logs.get('regression_loss'),'- train classification loss = ', logs.get('classification_loss'))
        print('val loss = ',self.history['total_val_loss'][-1])
        print('val regression loss = ',logs.get('val_regression_loss'),'- val classification loss = ', logs.get('val_classification_loss'))

**Comment by SeaBroNoMeal** [comment]
> I couldn't find the bug. But the issue is only with validation loss. So, it won't affect the training.
> But you can just add a callback in which you add the val_regression_loss and val_classification_loss and print that.
> 
> class LossHistory(keras.callbacks.Callback):
> 
> ```
> def __init__(self, verbose):
>     self.verbose = verbose
>     super(LossHistory, self).__init__()
> 
> def on_train_begin(self, logs={}):
>     self.history = {'loss':[], 'regression_loss':[], 'classification_loss':[],
>                     'total_val_loss':[], 'val_regression_loss':[], 'val_classification_loss':[]}
>     
> def on_epoch_end(self, epoch, logs={}):
>     
>     
>     print("Epoch - ",epoch+1)
>     
>     self.history['regression_loss'].append(logs.get('regression_loss'))
>     self.history['classification_loss'].append(logs.get('classification_loss'))
>     self.history['loss'].append(logs.get('loss'))
>     
>     self.history['val_regression_loss'].append(logs.get('val_regression_loss'))
>     self.history['val_classification_loss'].append(logs.get('val_classification_loss'))
>     self.history['total_val_loss'].append(logs.get('val_regression_loss')+logs.get('val_classification_loss'))
>     
>     logs['total_val_loss'] = self.history['total_val_loss'][-1]
>     
>     print('train loss = ',logs.get('loss'))
>     print('train regression loss = ',logs.get('regression_loss'),'- train classification loss = ', logs.get('classification_loss'))
>     print('val loss = ',self.history['total_val_loss'][-1])
>     print('val regression loss = ',logs.get('val_regression_loss'),'- val classification loss = ', logs.get('val_classification_loss'))
> ```

Hi, thanks for replying.

Could you please start a PR so I can see exactly where you put these codes? Also, if I do not add this class, can I just add up `val_classification_loss` and `val_regression_loss` to get the final `val_loss`, for instance for the case I provided above, can I add up 0.4772 and 0.6628 to get the true validation losee?

**Comment by ChunmianLin** [comment]
> > I couldn't find the bug. But the issue is only with validation loss. So, it won't affect the training.
> > But you can just add a callback in which you add the val_regression_loss and val_classification_loss and print that.
> > class LossHistory(keras.callbacks.Callback):
> > ```
> > def __init__(self, verbose):
> >     self.verbose = verbose
> >     super(LossHistory, self).__init__()
> > 
> > def on_train_begin(self, logs={}):
> >     self.history = {'loss':[], 'regression_loss':[], 'classification_loss':[],
> >                     'total_val_loss':[], 'val_regression_loss':[], 'val_classification_loss':[]}
> >     
> > def on_epoch_end(self, epoch, logs={}):
> >     
> >     
> >     print("Epoch - ",epoch+1)
> >     
> >     self.history['regression_loss'].append(logs.get('regression_loss'))
> >     self.history['classification_loss'].append(logs.get('classification_loss'))
> >     self.history['loss'].append(logs.get('loss'))
> >     
> >     self.history['val_regression_loss'].append(logs.get('val_regression_loss'))
> >     self.history['val_classification_loss'].append(logs.get('val_classification_loss'))
> >     self.history['total_val_loss'].append(logs.get('val_regression_loss')+logs.get('val_classification_loss'))
> >     
> >     logs['total_val_loss'] = self.history['total_val_loss'][-1]
> >     
> >     print('train loss = ',logs.get('loss'))
> >     print('train regression loss = ',logs.get('regression_loss'),'- train classification loss = ', logs.get('classification_loss'))
> >     print('val loss = ',self.history['total_val_loss'][-1])
> >     print('val regression loss = ',logs.get('val_regression_loss'),'- val classification loss = ', logs.get('val_classification_loss'))
> > ```
> 
> Hi, thanks for replying.
> 
> Could you please start a PR so I can see exactly where you put these codes? Also, if I do not add this class, can I just add up `val_classification_loss` and `val_regression_loss` to get the final `val_loss`, for instance for the case I provided above, can I add up 0.4772 and 0.6628 to get the true validation losee?

Hi, in your opinion, val_loss is not regression_loss plus classification_loss?
In mathematically, train_loss is regression_loss and classification loss, for instance, 1.2900 =0.6755 +0.6145. But why val_loss cannot calculate by the way? I am confused about this, do you understand?

Thanks!

**Comment by yashmarathe21** [comment]
> > > I couldn't find the bug. But the issue is only with validation loss. So, it won't affect the training.
> > > But you can just add a callback in which you add the val_regression_loss and val_classification_loss and print that.
> > > class LossHistory(keras.callbacks.Callback):
> > > ```
> > > def __init__(self, verbose):
> > >     self.verbose = verbose
> > >     super(LossHistory, self).__init__()
> > > 
> > > def on_train_begin(self, logs={}):
> > >     self.history = {'loss':[], 'regression_loss':[], 'classification_loss':[],
> > >                     'total_val_loss':[], 'val_regression_loss':[], 'val_classification_loss':[]}
> > >     
> > > def on_epoch_end(self, epoch, logs={}):
> > >     
> > >     
> > >     print("Epoch - ",epoch+1)
> > >     
> > >     self.history['regression_loss'].append(logs.get('regression_loss'))
> > >     self.history['classification_loss'].append(logs.get('classification_loss'))
> > >     self.history['loss'].append(logs.get('loss'))
> > >     
> > >     self.history['val_regression_loss'].append(logs.get('val_regression_loss'))
> > >     self.history['val_classification_loss'].append(logs.get('val_classification_loss'))
> > >     self.history['total_val_loss'].append(logs.get('val_regression_loss')+logs.get('val_classification_loss'))
> > >     
> > >     logs['total_val_loss'] = self.history['total_val_loss'][-1]
> > >     
> > >     print('train loss = ',logs.get('loss'))
> > >     print('train regression loss = ',logs.get('regression_loss'),'- train classification loss = ', logs.get('classification_loss'))
> > >     print('val loss = ',self.history['total_val_loss'][-1])
> > >     print('val regression loss = ',logs.get('val_regression_loss'),'- val classification loss = ', logs.get('val_classification_loss'))
> > > ```
> > 
> > 
> > Hi, thanks for replying.
> > Could you please start a PR so I can see exactly where you put these codes? Also, if I do not add this class, can I just add up `val_classification_loss` and `val_regression_loss` to get the final `val_loss`, for instance for the case I provided above, can I add up 0.4772 and 0.6628 to get the true validation losee?
> 
> Hi, in your opinion, val_loss is not regression_loss plus classification_loss?
> In mathematically, train_loss is regression_loss and classification loss, for instance, 1.2900 =0.6755 +0.6145. But why val_loss cannot calculate by the way? I am confused about this, do you understand?
> 
> Thanks!

No, val_loss is regression_loss plus classification loss. During the training, I was getting extremely low val_loss. I thought it was a bug in this repository but it has something to do with Keras. Since I can't find that bug and the training_loss is train_classification_loss plus train_regression_loss (so it isn't affecting the training process), I thought I can just write a custom callback where I will add the validation regression and classification losses. It doesn't solve the bug but it isn't affecting the training anyway. Using this callback, we just print the addition of the validation losses and maybe use this to save the model checkpoint based on this loss.

**Comment by yashmarathe21** [comment]
> > I couldn't find the bug. But the issue is only with validation loss. So, it won't affect the training.
> > But you can just add a callback in which you add the val_regression_loss and val_classification_loss and print that.
> > class LossHistory(keras.callbacks.Callback):
> > ```
> > def __init__(self, verbose):
> >     self.verbose = verbose
> >     super(LossHistory, self).__init__()
> > 
> > def on_train_begin(self, logs={}):
> >     self.history = {'loss':[], 'regression_loss':[], 'classification_loss':[],
> >                     'total_val_loss':[], 'val_regression_loss':[], 'val_classification_loss':[]}
> >     
> > def on_epoch_end(self, epoch, logs={}):
> >     
> >     
> >     print("Epoch - ",epoch+1)
> >     
> >     self.history['regression_loss'].append(logs.get('regression_loss'))
> >     self.history['classification_loss'].append(logs.get('classification_loss'))
> >     self.history['loss'].append(logs.get('loss'))
> >     
> >     self.history['val_regression_loss'].append(logs.get('val_regression_loss'))
> >     self.history['val_classification_loss'].append(logs.get('val_classification_loss'))
> >     self.history['total_val_loss'].append(logs.get('val_regression_loss')+logs.get('val_classification_loss'))
> >     
> >     logs['total_val_loss'] = self.history['total_val_loss'][-1]
> >     
> >     print('train loss = ',logs.get('loss'))
> >     print('train regression loss = ',logs.get('regression_loss'),'- train classification loss = ', logs.get('classification_loss'))
> >     print('val loss = ',self.history['total_val_loss'][-1])
> >     print('val regression loss = ',logs.get('val_regression_loss'),'- val classification loss = ', logs.get('val_classification_loss'))
> > ```
> 
> Hi, thanks for replying.
> 
> Could you please start a PR so I can see exactly where you put these codes? Also, if I do not add this class, can I just add up `val_classification_loss` and `val_regression_loss` to get the final `val_loss`, for instance for the case I provided above, can I add up 0.4772 and 0.6628 to get the true validation losee?

Yeah sure. You can add this class during the training process. Just add this custom callback during training the model and it will print val_loss and val_classificaiton_loss and val_regression_loss. It won't change this internally. I checked but this bug is related to Keras and not this repository. Maybe you can write your own custom function to train the model but its too much unnecessary hassle. Until your train_loss is the exact addition of train_classification_loss and train_regression_loss, you don't have to worry because your training process is going fine.

**Comment by RxstydnR** [comment]
FYI, in case, I share my same situation and solution.
I got the same problem, when I set validation data like below.
```python
history = model.fit(
            ...
            validation_data=[X_test, y_test],
            ...
        )
```
and solved when I use tuple like,

```python
history = model.fit(
            ...
            validation_data=(X_test, y_test),
            ...
        )

```
